### PR TITLE
feat(fluvio-cli): add base offset to partition list command

### PR DESCRIPTION
**Pull request by PanGan21**
Adds `base offset` to `fluvio partition list` command.
Related: #3975

### CODE REVIEW

File: crates/fluvio-controlplane/src/sc_api/update_lrs.rs [hunk 1]
```diff
@@ -62,6 +62,7 @@ pub struct LrsRequest {
     pub leader: ReplicaStatus,
     pub replicas: Vec<ReplicaStatus>,
     pub size: i64,
+    pub base_offset: i64,
```
Comment (morenol) [inline crates/fluvio-controlplane/src/sc_api/update_lrs.rs:65]: Try something like this
```suggestion
   #[fluvio(min_version = 1)]
    pub base_offset: i64,
```

and in the trait impl where we have:

```
impl Request for UpdateLrsRequest {

}
```
Add this: 
```
const DEFAULT_API_VERSION: i16 = 1;
```

File: crates/fluvio-controlplane-metadata/src/partition/status.rs [hunk 2]
```diff
@@ -38,6 +38,7 @@ pub struct PartitionStatus {
     #[fluvio(min_version = 5)]
     pub size: i64,
     pub is_being_deleted: bool,
+    pub base_offset: i64,
```
Comment (morenol) [inline crates/fluvio-controlplane-metadata/src/partition/status.rs:41]: Same here:
 ```suggestion
    #[fluvio(min_version = 15)]
    pub base_offset: i64,
```

And bump version of `impl Request for ObjectApiListRequest`

File: crates/fluvio-controlplane/src/sc_api/update_lrs.rs (hunk 1 above)
Comment (PanGan21) [reply crates/fluvio-controlplane/src/sc_api/update_lrs.rs:67]: Done!

File: crates/fluvio-controlplane-metadata/src/partition/status.rs (hunk 2 above)
Comment (PanGan21) [reply crates/fluvio-controlplane-metadata/src/partition/status.rs:42]: Done! Thanks for the help!

**Comment by digikata** [comment]
@EstebanBorai could you take a look at this https://github.com/infinyon/fluvio/actions/runs/9438543985/job/25995712397?pr=4059.  Offhand this may be from the test asserts needing to catch up to the optimized `fvm update` change

**Comment by EstebanBorai** [comment]
> @EstebanBorai could you take a look at this https://github.com/infinyon/fluvio/actions/runs/9438543985/job/25995712397?pr=4059. Offhand this may be from the test asserts needing to catch up to the optimized `fvm update` change

Giving it a check!

**Comment by EstebanBorai** [comment]
> > @EstebanBorai could you take a look at this https://github.com/infinyon/fluvio/actions/runs/9438543985/job/25995712397?pr=4059. Offhand this may be from the test asserts needing to catch up to the optimized `fvm update` change
> 
> Giving it a check!

Update! Here Im working out a fix to take in consideration the same version tests:
https://github.com/infinyon/fluvio/pull/4062

**Comment by PanGan21** [comment]
Hey @digikata  thanks for the assistance already.
Do you maybe know why upgrading doesn't work according to this pipeline?
https://github.com/infinyon/fluvio/actions/runs/9456228110/job/26051451364?pr=4059

File: crates/fluvio-controlplane/src/sc_api/update_lrs.rs (hunk 1 above)
Comment (EstebanBorai) [reply crates/fluvio-controlplane/src/sc_api/update_lrs.rs:67]: Given that we used `#[fluvio(min_version = 15)]` on `PartitionStatus` https://github.com/infinyon/fluvio/pull/4059/files#diff-f1e14778be799260d11d5529f1a8c3a5796cf26e4a8993eab058408b7b828a08L26, I think we should use the same value `15` for `pub base_offset: i64,` here?

**Review by EstebanBorai — COMMENTED** [review]
Good progress @PanGan21! Thanks again!

Can you try updating the version on `LrsRequest` to be `15` instead of `1` for `base_offset`?

File: crates/fluvio-controlplane/src/sc_api/update_lrs.rs (hunk 1 above)
Comment (EstebanBorai) [reply crates/fluvio-controlplane/src/sc_api/update_lrs.rs:67]: This way we also match against the new min version:
https://github.com/infinyon/fluvio/pull/4059/files#diff-b7ec42b71799d984a63822863daca2ce587a551f59133cafd64410e667df49c4R16

Comment (PanGan21) [reply crates/fluvio-controlplane/src/sc_api/update_lrs.rs:67]: Did that but it still seems that the upgrade is not correct.
```
Validate test data w/ v0.11.10-dev-1 CLI matches expected data AFTER upgrading cluster + CLI to v0.11.10-dev-1
shasum: WARNING: 1 computed checksum did NOT match
-: FAILED
Got: 
Expected: 22e48620de3e901be2d32b33c5f4430a67fc64ba
```

Can it be the global version?

**Comment by digikata** [comment]
Looking at the code again, I suspect `base_offset` needs to update `k8-util/helm/fluvio-sys/templates/crd_partition.yaml`  - at least in the status `additionalPrinterColumns` section.  Though I'm not sure this is the fault.

 @PanGan21 or @EstebanBorai, one way to try to get the error under a better analysis environment is to setup a k8s with stable fluvio and upgrade it. Otherwise most of this help is a shot in the dark. (Unfortunately, I don't think I have time to do that until next week).

**Comment by EstebanBorai** [comment]
> Looking at the code again, I suspect `base_offset` needs to update `k8-util/helm/fluvio-sys/templates/crd_partition.yaml` - at least in the status `additionalPrinterColumns` section. Though I'm not sure this is the fault.
> 
> @PanGan21 or @EstebanBorai, one way to try to get the error under a better analysis environment is to setup a k8s with stable fluvio and upgrade it. Otherwise most of this help is a shot in the dark. (Unfortunately, I don't think I have time to do that until next week).

Thanks for the recommendation @digikata! I will give this a shot and get back to this PR with news.

**Comment by EstebanBorai** [comment]
Hi @PanGan21! Im taking a look at this right now.

**Comment by PanGan21** [comment]
> PartitionSpec

Thanks for taking a look @EstebanBorai . Since we added base offset on the `PartitionStatus` I am wondering why it should also be in `PartitionSpec`. Also how did you manage to find this log? By upgrading a local cluster?

**Comment by EstebanBorai** [comment]
> > PartitionSpec
> 
> 
> 
> Thanks for taking a look @EstebanBorai . Since we added base offset on the `PartitionStatus` I am wondering why it should also be in `PartitionSpec`. Also how did you manage to get this log? By upgrading a local cluster?
> 
> 
> 
> From your experiment on #4077 I also see that I am probably missing `#[cfg_attr(feature = "use_serde", serde(default))]` on the base_offset. Should I add this as well?

Hi @PanGan21! 

You are right, thats an approach to fix the missing `baseOffset` error when parsing previous cluster CRDs on cluster upgrade.

File: k8-util/helm/fluvio-sys/templates/crd_partition.yaml [hunk 3]
```diff
@@ -143,6 +143,11 @@ spec:
         format: int64
         description: Leader High Watermark
         jsonPath: .status.leader.hw
+      - name: Base Offset
+        type: integer
+        format: int64
+        description: Base Offset
+        jsonPath: .status.baseOffset
       - name: EOF
```
Comment (sehz) [inline k8-util/helm/fluvio-sys/templates/crd_partition.yaml:151]: need to bump up helm chart version whenever there is change

Comment (PanGan21) [reply k8-util/helm/fluvio-sys/templates/crd_partition.yaml:151]: This value right?
https://github.com/infinyon/fluvio/blob/a5920e3d66fa5c8e6cbf4028542a3069504fdf70/k8-util/helm/fluvio-sys/templates/crd_partition.yaml#L12-L13

Comment (morenol) [reply k8-util/helm/fluvio-sys/templates/crd_partition.yaml:151]: Here https://github.com/infinyon/fluvio/blob/a5920e3d66fa5c8e6cbf4028542a3069504fdf70/k8-util/helm/fluvio-sys/Chart.yaml#L5

**Comment by EstebanBorai** [comment]
@PanGan21 Im working on state update improvements so we can land your PR!

Will get back to you in this PR when these changes are applied!

**Comment by EstebanBorai** [comment]
Hi @PanGan21! Can you please update your branch against `master`? We are close to merge this PR!

**Review by EstebanBorai — APPROVED** [review]

**Comment by PanGan21** [comment]
> Hi @PanGan21! Can you please update your branch against `master`? We are close to merge this PR!

I think there are still some issues in the pipeline unfortunately

**Review by sehz — APPROVED** [review]

**Comment by EstebanBorai** [comment]
Hi @PanGan21! Can you please rebase to have Rust 1.80 updates applied please?

**Comment by EstebanBorai** [comment]
Hi @PanGan21! Looks like the encoding version must be incremented as it conflicts with other
fields now.

I did an experiment here: https://github.com/infinyon/fluvio/pull/4119/commits/7a7455756b0aa4cb342950d290016dc331546534

Feel free to port these changes here!

File: crates/fluvio-controlplane-metadata/src/partition/status.rs [hunk 4]
```diff
@@ -38,6 +38,9 @@ pub struct PartitionStatus {
     #[fluvio(min_version = 5)]
     pub size: i64,
     pub is_being_deleted: bool,
+    #[cfg_attr(feature = "use_serde", serde(default))]
+    #[fluvio(min_version = 15)]
```
Comment (EstebanBorai) [inline crates/fluvio-controlplane-metadata/src/partition/status.rs:42]:
```suggestion
    #[fluvio(min_version = 16)]
```

**Review by EstebanBorai — COMMENTED** [review]
Just missing the bump for `status`!

**Review by EstebanBorai — APPROVED** [review]
Great work @PanGan21! Thanks so much for your efforts on this PR, please keep coming!

**Pull request merged**